aiohomekit/controller/coap/connection.py

```diff
@@ -0,0 +1,971 @@
+
+import asyncio
+import struct
+import logging
+import tlv8
+import uuid
+
+from aiocoap import Context, Message, resource
+from aiocoap.numbers.codes import Code
+from cryptography.exceptions import InvalidTag
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
+from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
+from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
+
+from aiohomekit.crypto.hkdf import hkdf_derive
+from aiohomekit.crypto.srp import SrpClient
+
+
+# globals/enums
+logger = logging.getLogger(__name__)
+
+class HAP_PAIRING_ERRORS:
+    UNKNOWN = 1
+    AUTHENTICATION = 2
+    MAX_PEERS = 4
+    MAX_TRIES = 5
+    UNAVAILABLE = 6
+    BUSY = 7
+
+class HAP_PAIRING_METHODS:
+    PAIR_SETUP = 0
+    PAIR_SETUP_WITH_AUTH = 1
+    PAIR_VERIFY = 2
+
+class HAP_PDU_OPCODES:
+    HAP_CHARACTERISTIC_SIGNATURE_READ = 0x01
+    HAP_CHARACTERISTIC_WRITE = 0x02
+    HAP_CHARACTERISTIC_READ = 0x03
+    HAP_CHARACTERISTIC_TIMEDWRITE = 0x04
+    HAP_CHARACTERISTIC_TIMEDREAD = 0x05
+    HAP_SERVICE_SIGNATURE_READ = 0x06
+    HAP_CHARACTERISTIC_CONFIGURATION = 0x07
+    HAP_PROTOCOL_CONFIGURATION = 0x08
+    UNK_09_READ_GATT = 0x09
+    UNK_0A = 0x0A
+    UNK_0B_SUBSCRIBE = 0x0B
+    UNK_0C = 0x0C
+    HAP_TOKEN_REQUEST = 0x10
+    HAP_TOKEN_UPDATE_REQUEST = 0x11
+    HAP_INFO_REQUEST = 0x12
+
+class HAP_PDU_TLV_TAGS:
+    SEPARATOR = 0x00
+    HAP_PARAM_VALUE = 0x01
+    HAP_PARAM_ADDITIONAL_AUTHORIZATION_DATA = 0x02
+    HAP_PARAM_ORIGIN = 0x03
+    HAP_PARAM_CHARACTERISTIC_TYPE = 0x04
+    HAP_PARAM_CHARACTERISTIC_INSTANCE_ID = 0x05
+    HAP_PARAM_SERVICE_TYPE = 0x06
+    HAP_PARAM_SERVICE_INSTANCE_ID = 0x07
+    HAP_PARAM_TTL = 0x08
+    HAP_PARAM_RETURN_RESPONSE = 0x09
+    HAP_PARAM_HAP_CHARACTERISTIC_PROPERTIES_DESCRIPTOR = 0x0A
+    HAP_PARAM_GATT_USER_DESCRIPTION_DESCRIPTOR = 0x0B
+    HAP_PARAM_GATT_PRESENTATION_FORMAT_DESCRIPTOR = 0x0C
+    HAP_PARAM_GATT_VALID_RANGE = 0x0D
+    HAP_PARAM_HAP_STEP_VALUE_DESCRIPTOR = 0x0E
+    HAP_PARAM_HAP_SERVICE_PROPERTIES = 0x0F
+    HAP_PARAM_HAP_LINKED_SERVICES = 0x10
+    HAP_PARAM_HAP_VALID_VALUES_DESCRIPTOR = 0x11
+    HAP_PARAM_HAP_VALID_VALUES_RANGE_DESCRIPTOR = 0x12
+    UNK_13_CHARACTERISTIC = 0x13
+    UNK_14_CHARACTERISTICS = 0x14
+    UNK_15_SERVICE = 0x15
+    UNK_16_SERVICES = 0x16
+    UNK_17 = 0x17
+    UNK_18 = 0x18
+    UNK_19 = 0x19
+    UNK_1A_ACCESSORY_INSTANCE_ID = 0x1A
+
+class HAP_TLV_TAGS:
```

Comment:
Some crossover here with https://github.com/Jc2k/aiohomekit/blob/5902621ca0a1c8bd4399fcf8f2b7767d880cc252/aiohomekit/protocol/tlv.py#L44.

Reply:
Now using an IntEnum in tlv.py for HAP TLV tags.